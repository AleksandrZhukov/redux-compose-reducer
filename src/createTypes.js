import {
  namespacedActionType,
  createTypesObject,
  createCapturedError,
} from './utils'

const createTypes = (namespace, types = []) => {
  if (
    typeof namespace !== 'string' ||
    namespace.length === 0 ||
    !Array.isArray(types) ||
    types.length === 0
  )
    throw createError(ARGUMENT_ERROR)

  const result = createTypesObject()
  types.forEach(type => {
    result[type] = namespacedActionType(namespace, type)
  })
  return result
}

const createError = message => createCapturedError(message, createTypes)

export const ARGUMENT_ERROR =
  'Expected first argument (namespace) to be a non empty string ' +
  'and second argument (types) to be a non empty array.'

export default createTypes
