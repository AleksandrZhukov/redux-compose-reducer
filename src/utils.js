const DELIMITER = '/'

export const namespacedActionType = (namespace, type) =>
  `${namespace}${DELIMITER}${type}`

export const isObject = o => o instanceof Object && !Array.isArray(o)

export const createCapturedError = (message, cst) => {
  const error = new Error(message)
  if (cst && Error.captureStackTrace) Error.captureStackTrace(error, cst)
  return error
}

const TYPES_PROXY_HANDLER = {
  get(target, code) {
    if (code in target) {
      return target[code]
    }

    const keys = Object.keys(target).join(', ')
    throw createCapturedError(`Key '${code}' does not exist.`, this.get)
  },
}

export const createTypesObject = () => {
  const useProxy =
    process.env.NODE_ENV !== 'production' &&
    'Proxy' in global &&
    typeof Proxy === 'function'

  return useProxy ? new Proxy({}, TYPES_PROXY_HANDLER) : {}
}
