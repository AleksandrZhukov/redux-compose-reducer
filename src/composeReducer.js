import { namespacedActionType, isObject, createCapturedError } from './utils'

const composeReducer = (...args) => {
  const {
    namespace,
    types,
    initialState,
    reducers = {},
    globalReducer,
  } = prepareParams(args)

  if (
    (types !== undefined && !isObject(types)) ||
    (types === undefined &&
      (typeof namespace !== 'string' || namespace.length === 0)) ||
    !isObject(initialState) ||
    !isObject(reducers) ||
    (globalReducer !== undefined && typeof globalReducer !== 'function')
  )
    throw createError(ARGUMENT_ERROR)

  const typeToReducerMap = types
    ? createMapFromTypes(types, reducers)
    : createMapFromNamespace(namespace, reducers)

  return (state = initialState, action) => {
    state = typeToReducerMap[action.type]
      ? typeToReducerMap[action.type](state, action)
      : state
    return globalReducer ? globalReducer(state, action) : state
  }
}

const prepareParams = args => {
  if (typeof args[0] === 'string') {
    console.warn(DEPRICATED_API_WARNING)
    return {
      namespace: args[0],
      reducers: args[1],
      initialState: args[2],
      globalReducer: args[3],
    }
  }

  return args[0] || {}
}

const createMapFromTypes = (types, reducers) => {
  const result = {}
  Object.keys(reducers).forEach(key => {
    if (!(key in types)) throw createError(`There is no '${key}' action type.`)
    result[types[key]] = reducers[key]
  })
  return result
}

const createMapFromNamespace = (namespace, reducers) => {
  const result = {}
  Object.keys(reducers).forEach(type => {
    result[namespacedActionType(namespace, type)] = reducers[type]
  })
  return result
}

export const ARGUMENT_ERROR = `As argument expected object of shape : {
  namespace: 'non empty string',
  types: 'object',
  initialState: 'object',
  reducers: 'object',
  globalReducer: 'function'
}.
Required keys: namespace or types, initialState.`

const DEPRICATED_API_WARNING =
  'redux-compose-reducer#composeReducer: ' +
  'Multiple arguments api is depricated and will be removed in future versions. ' +
  'Please use object argument.'

const createError = message => createCapturedError(message, composeReducer)

export default composeReducer
