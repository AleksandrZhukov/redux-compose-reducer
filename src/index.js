import createTypes from './createTypes'
import composeReducer from './composeReducer'

export { createTypes, composeReducer }
