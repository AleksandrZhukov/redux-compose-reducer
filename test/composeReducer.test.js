import * as utils from '../src/utils'
import { composeReducer, createTypes } from '../src'
import { ARGUMENT_ERROR } from '../src/composeReducer'

describe('composeReducer', () => {
  beforeEach(() => {
    console.warn = jest.fn()
  })

  it('produces correct state depends on action type', () => {
    const reducer = composeReducer({
      namespace: 'test',
      initialState: { action: false, action2: false },
      reducers: {
        action: state => Object.assign({}, state, { action: true }),
        action2: state => Object.assign({}, state, { action2: true }),
      },
    })

    let actual = reducer(undefined, { type: 'test/action' })
    expect(actual).toEqual({
      action: true,
      action2: false,
    })

    actual = reducer(undefined, { type: 'test/action2' })
    expect(actual).toEqual({
      action: false,
      action2: true,
    })
  })

  it('ignores unknown actions', () => {
    const action = jest.fn()
    const reducer = composeReducer({
      namespace: 'test',
      initialState: { ok: false },
      reducers: { action },
    })

    const expected = { ok: false }
    const actual = reducer(undefined, { type: 'external/action' })
    expect(actual).toEqual(expected)
    expect(action.mock.calls.length).toBe(0)
  })

  describe('without namespace', () => {
    it('produces correct state depends on action type', () => {
      const reducer = composeReducer({
        initialState: { action: false, action2: false },
        reducers: {
          action: state => Object.assign({}, state, { action: true }),
          action2: state => Object.assign({}, state, { action2: true }),
        },
      })

      let actual = reducer(undefined, { type: 'action' })
      expect(actual).toEqual({
        action: true,
        action2: false,
      })

      actual = reducer(undefined, { type: 'action2' })
      expect(actual).toEqual({
        action: false,
        action2: true,
      })
    })
  })

  describe('`types` attribute', () => {
    it('produces correct state depends on action type', () => {
      const types = createTypes('test', ['action', 'action2'])
      const reducer = composeReducer({
        types,
        initialState: { action: false, action2: false },
        reducers: {
          action: state => Object.assign({}, state, { action: true }),
          action2: state => Object.assign({}, state, { action2: true }),
        },
      })

      let actual = reducer(undefined, { type: types.action })
      expect(actual).toEqual({
        action: true,
        action2: false,
      })

      actual = reducer(undefined, { type: types.action2 })
      expect(actual).toEqual({
        action: false,
        action2: true,
      })
    })

    it('validates reducers', () => {
      const types = createTypes('test', ['action', 'action2'])
      const creator = () =>
        composeReducer({
          types,
          initialState: { action: false, action2: false },
          reducers: {
            action: state => Object.assign({}, state, { action: true }),
            action3: state => Object.assign({}, state, { action2: true }),
          },
        })

      expect(creator).toThrowErrorMatchingSnapshot()
    })
  })

  describe('`reducers` shortcuts', () => {
    it('produces correct state depends on action type', () => {
      const reducer = composeReducer({
        namespace: 'test',
        initialState: { action: false, action2: false },
        reducers: {
          action: { action: true },
          action2: { action2: true },
        },
      })

      let actual = reducer(undefined, { type: 'test/action' })
      expect(actual).toEqual({
        action: true,
        action2: false,
      })

      actual = reducer(undefined, { type: 'test/action2' })
      expect(actual).toEqual({
        action: false,
        action2: true,
      })
    })
  })

  describe('`globalReducer` attribute', () => {
    it('allows to reduce global actions', () => {
      const GLOBAL_ACTION_TYPE = 'GLOBAL/ACTION'

      const globalReducer = (state, action) => {
        switch (action.type) {
          case GLOBAL_ACTION_TYPE:
            return { ok: true }
          default:
            return state
        }
      }
      const reducer = composeReducer({
        namespace: 'test',
        initialState: { ok: false },
        globalReducer,
      })

      const expected = { ok: true }
      const actual = reducer(undefined, { type: GLOBAL_ACTION_TYPE })
      expect(actual).toEqual(expected)
    })

    it('works after `reducers` attribute', () => {
      const globalReducer = (state, action) => {
        switch (action.type) {
          case 'test/action':
            expect(state).toEqual({ ok: true })
            return state

          default:
            return state
        }
      }
      const reducer = composeReducer({
        namespace: 'test',
        initialState: { ok: false },
        reducers: {
          action: () => ({ ok: true }),
        },
        globalReducer,
      })

      const expected = { ok: true }
      const actual = reducer(undefined, { type: 'test/action' })
      expect(actual).toEqual(expected)
    })
  })

  describe('depricated api', () => {
    it('still supported', () => {
      const reducer = composeReducer(
        'test',
        {
          action: () => ({ ok: true }),
        },
        { ok: false }
      )

      const expected = { ok: true }
      const actual = reducer(undefined, { type: 'test/action' })
      expect(actual).toEqual(expected)
    })

    it('produces console.warn', () => {
      console.warn = jest.fn()
      const reducer = composeReducer(
        'test',
        {
          action: () => ({ ok: true }),
        },
        { ok: false }
      )

      reducer(undefined, { type: 'test/action' })
      const { calls } = console.warn.mock
      expect(calls.length).toBe(1)
      expect(calls[0][0]).toMatchInlineSnapshot(
        `"redux-compose-reducer#composeReducer: Multiple arguments api is depricated and will be removed in future versions. Please use object argument."`
      )
    })
  })

  describe('validation', () => {
    describe('without parameters', () => {
      it('produces error', () => {
        expect(() => composeReducer()).toThrowError(ARGUMENT_ERROR)
      })
    })

    describe('with incorrect namespace attribute', () => {
      it('produces error', () => {
        expect(() => composeReducer({ namespace: {} })).toThrowError(
          ARGUMENT_ERROR
        )
      })
    })

    describe('with incorrect types attribute', () => {
      it('produces error', () => {
        expect(() => composeReducer({ types: null })).toThrowError(
          ARGUMENT_ERROR
        )
      })
    })

    describe('with incorrect initialState attribute', () => {
      it('produces error', () => {
        expect(() =>
          composeReducer({ namespace: 'test', initialState: false })
        ).toThrowError(ARGUMENT_ERROR)
      })
    })

    describe('with incorrect reducers attribute', () => {
      it('produces error', () => {
        expect(() =>
          composeReducer({
            namespace: 'test',
            initialState: {},
            reducers: null,
          })
        ).toThrowError(ARGUMENT_ERROR)
      })
    })

    describe('with incorrect globalReducer attribute', () => {
      it('produces error', () => {
        expect(() =>
          composeReducer({
            namespace: 'test',
            initialState: {},
            globalReducer: {},
          })
        ).toThrowError(ARGUMENT_ERROR)
      })
    })
  })

  describe('unknown keys', () => {
    const reducer = composeReducer({
      namespace: 'test',
      initialState: {
        knownAttribute: true,
      },
      reducers: {
        action: state => Object.assign({}, state, { unknownAttribute: true }),
      },
    })

    afterEach(() => {
      utils.isDev = true
    })

    describe('development build', () => {
      it('produces warning', () => {
        reducer(undefined, { type: 'test/action' })

        const { calls } = console.warn.mock
        expect(calls.length).toBe(1)
        expect(calls[0][0]).toMatchInlineSnapshot(
          `"Reducer for action 'test/action' produced new keys: [unknownAttribute]"`
        )
      })
    })

    describe('production build', () => {
      it('works silently', () => {
        utils.isDev = false
        reducer(undefined, { type: 'test/action' })
        expect(console.warn.mock.calls.length).toBe(0)
      })
    })
  })
})
