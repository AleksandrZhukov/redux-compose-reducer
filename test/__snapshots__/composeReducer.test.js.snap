// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composeReducer \`types\` attribute validates reducers 1`] = `"There is no 'action3' action type."`;
